File: 003_Longest_Substring_Without_Repeating_Characters/solution.go
func lengthOfLongestSubstring(s string) int {
    byteIndexMap := map[byte]int{}
    max := 0
    for i := range s {
        _, ok := byteIndexMap[s[i]]
        if !ok {
            byteIndexMap[s[i]] = i
            if len(byteIndexMap) > max {
                max = len(byteIndexMap)
            }
        } else {
            old := byteIndexMap[s[i]]
            for key, value := range byteIndexMap {
                if value < old {
                    delete(byteIndexMap, key)
                }
            }
            byteIndexMap[s[i]] = i  
        }
    }
    return max
}

File: 003_Longest_Substring_Without_Repeating_Characters/solution.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0
        j = 0
        max = 0
        exist = {}
        for n in range(len(s)):
            exist[s[n]] = -1
        while j < len(s):
            if exist[s[j]] == -1:
                exist[s[j]] = 1
                j += 1
            else:
                while s[i] != s[j]:
                    exist[s[i]] = -1
                    i += 1
                i += 1
                j += 1
            max = max if max > (j - i) else (j - i)
        return max
